src/resolve/fuzz-resource-record.c

```diff
@@ -28,8 +29,11 @@ int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size) {
 
         assert_se(dns_resource_record_to_json(rr, &v) >= 0);
         assert_se(sd_json_variant_dump(v, SD_JSON_FORMAT_PRETTY|SD_JSON_FORMAT_COLOR|SD_JSON_FORMAT_SOURCE, f, NULL) >= 0);
-        assert_se(dns_resource_record_to_wire_format(rr, false) >= 0);
-        assert_se(dns_resource_record_to_wire_format(rr, true) >= 0);
+
+        r = dns_resource_record_to_wire_format(rr, /* canonical= */ false);
+        assert_se(r >= 0 || r == -EINVAL);
```

Comment:
I'm not sure where `-EINVAL` can come from here. It would mean that `dns_resource_record_new_from_raw` accepts and pass on resource records it shouldn't accept. Does it actually crash if this commit is reverted?

Reply:
Looks like it actually crashes if `-EINVAL` is removed. Before I forget here is what triggers the crash:
```python
>>> b
b"???????????????@?\x1c\x02\xf3\x02.'\x02\x02\x02\x02/\x02\x02\x01'\x1b\x80\x00\x00\x00\x00\x00\x00\x01\x01\x01\x02\x13\x01\x01\x01\x01\xff\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x00\x00\x00\x01\x07\x00\x00\x00\x00\x00\x00"
>>> DNSRR(b).show()
###[ DNS Resource Record ]###
  rrname    = b"??????????????@?\x1c\x02\xf3\x02.'\x02\x02\x02\x02/\x02\x02\x01'\x1b\x80\x00\x00\x00\x00\x00\x00\x01\x01\x01\x02\x13\x01\x01\x01\x01\xff\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01.\x01."
  type      = RESERVED
  cacheflush= 0
  rclass    = 263
  ttl       = 0
  rdlen     = 0
  rdata     = b''
```
`tshark` shows
```sh
    Answers
        ??????????????@?\034\002�\002.'\002\002\002\002/\002\002\001'\033�\000\000\000\000\000\000\001\001\001\002\023\001\001\001\001�\001\001\001\001\001\001\001\001\001\001\001\001\001\001.\001: type Unused, class Unknown
            Name: ??????????????@?\034\002�\002.'\002\002\002\002/\002\002\001'\033�\000\000\000\000\000\000\001\001\001\002\023\001\001\001\001�\001\001\001\001\001\001\001\001\001\001\001\001\001\001.\001
            Type: Unused (0) 
            Class: Unknown (0x0107)
            Time to live: 0 (0 seconds)
            Data length: 0
```

Reply:
So the issue here is that the parser thinks that the length of the two labels exceeds `DNS_HOSTNAME_MAX` (253) and that's completely wrong because the actual length of the first label is 63 and the actual length of the second label is 1 and their total length is far from `DNS_HOSTNAME_MAX`. The check is buggy because it looks at their lengths when they are escaped as opposed to their actual length in the wire format. I have no idea how it can be "good-to-merge".

Reply:
I opened https://github.com/systemd/systemd/issues/34399. (It would be nice not to change the fuzz target because it tries to make sure that whatever resolved does makes sense)

Reply:
As far as I can tell https://github.com/systemd/systemd/issues/34399 was closed so this thread can probably be resolved.